src/pymor/bindings/fenics.py

```diff
@@ -194,6 +194,21 @@ def _real_apply_inverse_one_vector(self, v, mu=None, initial_guess=None,
             _apply_inverse(self.matrix, r.impl, v.impl, options)
             return r
 
+        def _real_apply_inverse_adjoint_one_vector(self, u, mu=None, initial_guess=None,
+                                                   least_squares=False, prepare_data=None):
+            if least_squares:
+                raise NotImplementedError
+            r = (self.range.real_zero_vector() if initial_guess is None else
+                 initial_guess.copy(deep=True))
+            options = self.solver_options.get('inverse') if self.solver_options else None
+            if not hasattr(self, '_matrix_transpose'):
+                self._matrix_transpose = self.matrix.copy()
+                mat = df.as_backend_type(self.matrix).mat()
+                mat_tr = df.as_backend_type(self._matrix_transpose).mat()
+                mat.transpose(mat_tr)
```

Comment:
I would assume that this really creates a new sparse matrix storing the transpose of the original matrix? This does not sound like a good was to do this. Judging from [here](https://www.mcs.anl.gov/petsc/petsc4py-current/docs/apiref/petsc4py.PETSc.Mat-class.html) there should also be a `solveTranspose` method, which is probably a better way to do things.

Reply:
@sdrave as far as I see that is not properly wrapped in dolfin, so wouldn't that mean we have to restrict to `tensor=PETScMatrix` in all `assemble` calls to ensure that PETSc is actually used as the backend?

Reply:
> I would assume that this really creates a new sparse matrix storing the transpose of the original matrix? This does not sound like a good was to do this. Judging from [here](https://www.mcs.anl.gov/petsc/petsc4py-current/docs/apiref/petsc4py.PETSc.Mat-class.html) there should also be a `solveTranspose` method, which is probably a better way to do things.

I remember seeing `solveTranspose`, but I went with `transpose + df.solve` because `solveTranspose` doesn't have the options that `df.solve` has (e.g., choosing a solver and a precondition). If we go with `solveTranspose`, I guess `apply_inverse` should use PETSc's `solve` for consistency.

Reply:
I realize now that this is `FenicsMatrixOperator` not `FenicsOperator` so maybe my PETSc vs Dolfin comment is not as important?!

Reply:
I did a bit of research myself and, disappointingly, there really does not seem to be a good solution for this. Maybe, we for now stick with you implementation, @pmli, but output some warning in for the case that a iterative solve has been selected? (If you can afford a direct solver, then copying the matrix probably isn't so bad after all.) At least, we should have a code comment on why we are doing this.